provider/image.go

```diff
@@ -401,13 +402,140 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
+// 1. We label the image with a unique build ID.
+// 2. We poll the image store for the image ID we expect to find, and return that ID.
+// 3. We only return an imageId if the image in the store matches the image we built.
+func (p *dockerNativeProvider) runImageBuild(
+	ctx context.Context, docker *client.Client, tar io.Reader,
+	opts types.ImageBuildOptions, urn resource.URN, name string) (string, error) {
+	if opts.Labels == nil {
+		opts.Labels = make(map[string]string)
+	}
+	// TODO: https://github.com/pulumi/pulumi-docker/issues/846 - Consider removing the build ID.
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return "", fmt.Errorf("error generating random ID for build: %v", err)
+	}
+	buildIDLabel := "pulumi.com/build-id"
+	buildIDValue := id.String()
+	opts.Labels[buildIDLabel] = buildIDValue
+
+	// Close the imgBuildResp in a timely manner
+	_ = p.host.LogStatus(ctx, "info", urn, "Starting Docker build")
+	imgBuildResp, err := docker.ImageBuild(ctx, tar, opts)
+	if err != nil {
+		return "", fmt.Errorf("error building image: %v", err)
+	}
+	defer imgBuildResp.Body.Close()
+
+	var imageID string
+	extractImageID := func(rm json.RawMessage) (bool, string, error) {
+		var result types.BuildResult
+		err := json.Unmarshal(rm, &result)
+		if err != nil {
+			// Unmarshal failures here mean the message isn't what we expected, return handled = false
+			return false, "", nil
+		}
+
+		imageID = result.ID
+		return true, fmt.Sprintf("Built image with ID %s", imageID), nil
+	}
+
+	err = p.processLog(ctx, urn, imgBuildResp.Body, extractImageID)
+	if err != nil {
+		return "", fmt.Errorf("error reading build output: %v", err)
+	}
+
+	if imageID == "" {
+		return "", fmt.Errorf("no image id found in build output")
+	}
+
+	_ = p.host.LogStatus(ctx, "info", urn,
+		fmt.Sprintf("Built image with local id %q, polling image store for image", imageID))
+	// TODO: https://github.com/pulumi/pulumi-docker/issues/846 - Consider removing polling for the
+	// image.
+	findImageID := func() (bool, error) {
+		listResult, err := docker.ImageList(ctx, types.ImageListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("error inspecting image: %v", err)
+		}
+		// Search for our imageId in listResult
+		for _, storedImage := range listResult {
+			if storedImage.ID == imageID {
+				if storedImage.Labels[buildIDLabel] == buildIDValue {
+					return true, nil
+				}
+				_ = p.host.Log(ctx, "warning", urn,
+					fmt.Sprintf("Found in store does not match built image, expected label %s=%s, found %s",
+						buildIDLabel, buildIDValue, storedImage.Labels[buildIDLabel]))
+				return false, nil
+			}
+		}
+		return false, nil
+	}
+
+	for i := 0; i < 60; i++ {
+		found, err := findImageID()
+		if err != nil {
+			return "", err
+		}
+		if found {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	_ = p.host.LogStatus(ctx, "info", urn, fmt.Sprintf("Image built successfully, local id %q", imageID))
```

Comment:
Should this persist to final log output? Can go either way here.
```suggestion
	_ = p.host.Log(ctx, "info", urn, fmt.Sprintf("Image built successfully, local id %q", imageID))
```

Reply:
I think we should only log actionable messages to persistent output, avoid cluttering output.